### PR TITLE
improvement: Only reindex workspace on onBuildTargetChanged

### PR DESCRIPTION
**Pull request by tgodzik**
Previously, we would always reconnect to build server, which is not needed and takes longer. Instead now we only reimport and reindex the workspace.

Related to https://github.com/VirtusLab/scala-cli/issues/2412

### CODE REVIEW

File: tests/unit/src/test/scala/tests/BspBuildChangedSuite.scala [hunk 1]
```diff
@@ -1,61 +0,0 @@
-package tests
-
-import scala.concurrent.Promise
-
-import scala.meta.internal.metals.MetalsEnrichments._
-
-import bill.Bill
-
-class BspBuildChangedSuite extends BaseLspSuite("bsp-build-changed") {
```
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/BspBuildChangedSuite.scala:9]: So this suite was testing exactly what I removed. I think the Scala CLI test makes much more sense.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 2]
```diff
@@ -2008,38 +2018,32 @@ class MetalsLspService(
...
-                  .flatMap(_ => quickConnectToBuildServer())
-            case (Some(digest), _) if isBloopOrEmpty =>
-              slowConnectToBloopServer(forceImport, buildTool, digest)
-            case (Some(digest), _) =>
-              indexer.reloadWorkspaceAndIndex(
-                forceImport,
-                buildTool,
-                digest,
-                importBuild,
+        case Some(BuildTool.Found(buildTool: ScalaCliBuildTool, _))
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2021]: This is just a refactor to use BuildTool.Found with checksum ready.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 3]
```diff
@@ -2145,13 +2151,20 @@ class MetalsLspService(
...
+    if (otherChanges0.nonEmpty) {
+      bspSession match {
+        case None => scribe.warn("No build server connected")
+        case Some(session) =>
+          for {
+            _ <- importBuild(session)
+            _ <- indexer.profiledIndexWorkspace(runDoctorCheck)
+          } {
+            focusedDocument().foreach(path => compilations.compileFile(path))
+            compilers.cancel()
```
Comment (kasiaMarek) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2163]: `importBuild` does that too, why do we need this again?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 4]
```diff
@@ -2114,9 +2118,11 @@ class MetalsLspService(
       change
     }
 
-  private def maybeQuickConnectToBuildServer(
+  private def onBuildTargetChanges(
       params: b.DidChangeBuildTarget
   ): Unit = {
+    // Make sure that no compilation is running, if it is it might not get completed properly
```
Comment (kasiaMarek) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: How exactly?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: I am not sure, I could not reproduce it reliably. My guess is that when you send one compile first and then change using directives in another compile requests a short while later, last one will get cancelled and the first one forgotten.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala (hunk 3 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2163]: Hmm... it wasn't working without it, I think we might actually need to move it as last.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala (hunk 4 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: And get stuck in the queue until next cancel.

Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: So it's a `BatchedFunction` thing not a `scala-cli` thing?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: yes, but it showed up here multiple times when testing.

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2124]: Yeah, it seems Scala CLI doesn't always send the respond, looks like it's a race condition, but might be hard to fix.

**Review by kasiaMarek — APPROVED** [review]

**Comment by tgodzik** [comment]
Let's wait to merge it until after release to be on the safe side.

**Pull request merged**